kakao/src/main/kotlin/com/agoda/kakao/screen/Screen.kt

```diff
@@ -24,15 +27,104 @@ import com.agoda.kakao.common.views.KBaseView
 @Suppress("UNCHECKED_CAST")
 @KakaoDslMarker
 open class Screen<out T: Screen<T>>: ScreenActions {
-    override val view: ViewInteraction = Espresso.onView(ViewMatchers.isRoot())
-    operator fun invoke(function: T.() -> Unit) = function.invoke(this as T)
+    private var vi: Interceptor<ViewInteraction, ViewAssertion, ViewAction>? = null
+    private var di: Interceptor<DataInteraction, ViewAssertion, ViewAction>? = null
+    private var wi: Interceptor<Web.WebInteraction<*>, WebAssertion<*>, Atom<*>>? = null
+
+    override val view: ViewInteractionDelegate = ViewInteractionDelegate(Espresso.onView(ViewMatchers.isRoot()))
+
+    private var isPushed = false
 
     /**
      * The visibility of rootView will be checked when entering the screen
-     * @rootView.isVisible() Will be called after land onScreen<>()
+     * @rootView.isVisible() Will be called after land onScreen<>() and after invoking screen directly.
      */
     open var rootView: KBaseView<*>? = null
 
+    /**
+     * Sets the interceptors for the screen.
+     * Interceptors will be invoked on all interactions while the screen is active.
+     *
+     * The screen is considered `active` when it is invoked in one of the following ways:
+     * ```
+     * val screen = SomeScreen()
+     *
+     * screen { // Active
+     *     view { click() }
+     *     ...
+     * } // Inactive
+     *
+     * // OR
+     *
+     * onScreen<SomeScreen>() { // Active
+     *     view { click() }
+     *     ...
+     * } // Inactive
+     * ```
+     *
+     * If you use nesting, interceptors of the latest screen that became active will be invoked.
+     * All active screen's interceptors are being stored in a stack, so that you won't.
+     *
+     * @param configurator Configuration of the interceptors
+     *
+     * @see Interceptor
+     */
+    fun intercept(configurator: Interceptor.Configurator.() -> Unit) {
+        var push = false
+
+        if (isPushed) {
+            push = true
+            pop()
+        }
+
+        Interceptor.Configurator().apply(configurator).configure().also {
+            vi = it.first
+            di = it.second
+            wi = it.third
+        }
+
+        if (push) push()
+    }
+
+    /**
+     * Removes the interceptors from the screen.
+     *
+     * @see intercept
+     * @see Interceptor
+     */
+    fun reset() {
+        if (isPushed) pop()
+        vi = null
+        di = null
+        wi = null
+    }
+
+    /**
+     * Operator that allows usage of DSL style
+     *
+     * @param function Tail lambda with receiver which is your screen
+     */
+    operator fun invoke(function: T.() -> Unit) {
+        if (!isPushed) push()
+        rootView?.isVisible()
+        function.invoke(this as T)
+        if (isPushed) pop()
```

Comment:
Maybe it makes sense to set popped interceptors here to vi, di and wi for possible reuse?

Reply:
These popped interceptors were pushed to the stack directly from the properties in the first place.